rust/src/passwd.rs

```diff
@@ -174,16 +175,146 @@ pub fn migrate_group_except_root(rootfs_dfd: i32, preserved_groups: &Vec<String>
     Ok(())
 }
 
-/// Merge and deduplicate entries from /usr/etc and /usr/lib.
-pub fn concat_fs_content(
+pub fn passwd_compose_prep(rootfs_dfd: i32, treefile: &mut Treefile) -> Result<()> {
+    let rootfs = ffiutil::ffi_view_openat_dir(rootfs_dfd);
+    passwd_compose_prep_impl(&rootfs, treefile, None, true)
+}
+
+/// Passwd/group handler for composes/treefiles.
+///
+/// We support various passwd/group handling. This function is primarily
+/// responsible for handling the "previous" and "file" paths; in both
+/// cases we inject data into the tree before even laying
+/// down any files, and notably before running RPM `useradd` etc.
+pub fn passwd_compose_prep_repo(
     rootfs_dfd: i32,
+    treefile: &mut Treefile,
     mut ffi_repo: Pin<&mut crate::ffi::OstreeRepo>,
     previous_checksum: &str,
+    unified_core: bool,
 ) -> Result<()> {
-    anyhow::ensure!(!previous_checksum.is_empty(), "missing previous reference");
-
     let rootfs = ffiutil::ffi_view_openat_dir(rootfs_dfd);
     let repo = ffi_repo.gobj_wrap();
+    passwd_compose_prep_impl(
+        &rootfs,
+        treefile,
+        Some((&repo, previous_checksum)),
+        unified_core,
+    )
+}
+
+fn passwd_compose_prep_impl(
+    rootfs: &openat::Dir,
+    treefile: &mut Treefile,
+    repo_previous_rev: Option<(&ostree::Repo, &str)>,
```

Comment:
Neat.  I forget that syntax is supported.  Definitely useful in situations like this.